cpp/src/arrow/util/thread_pool.h

```diff
@@ -194,6 +196,11 @@ class ARROW_EXPORT Executor {
   // Executor. Returns false if this Executor does not support this property.
   virtual bool OwnsThisThread() { return false; }
 
+  // Return true if this is the current executor being called
+  // n.b. this defaults to just calling OwnsThisThread
+  // unless the threadpool is disabled
+  virtual bool IsCurrentExecutor() { return OwnsThisThread(); }
```

Comment:
It's because in threading disabled mode, there can be multiple nested executors running, with one executor waiting on another. OwnsThisThread assumes that each executor has a thread, and uses the thread id to check if the code is running under one particular executor, which doesn't work without threads.